src/providers/shapeshift/shapeshift.ts

```diff
@@ -162,13 +194,241 @@ export class ShapeshiftProvider {
+      this.credentials.CLIENT_ID +
+      '&redirect_uri=' +
+      this.credentials.REDIRECT_URI
+    );
+  }
+
+  public getSignupUrl() {
+    return this.credentials.HOST + '/signup';
+  }
+
+  public getStoredToken(cb) {
+    this.persistenceProvider
+      .getShapeshiftToken(this.credentials.NETWORK)
+      .then(accessToken => {
+        if (!accessToken) return cb();
+        return cb(accessToken);
+      })
+      .catch(_ => {
+        return cb();
+      });
+  }
+
+  public getToken(code, cb) {
+    let url = this.credentials.HOST + '/oauth/token';
+    let data = {
+      grant_type: 'authorization_code',
+      code,
+      client_id: this.credentials.CLIENT_ID,
+      client_secret: this.credentials.CLIENT_SECRET,
+      redirect_uri: this.credentials.REDIRECT_URI
+    };
+    const headers = new HttpHeaders({
+      'Content-Type': 'application/json',
+      Accept: 'application/json'
+    });
+
+    this.http.post(url, data, { headers }).subscribe(
+      data => {
+        this.logger.info('ShapeShift: GET Access Token: SUCCESS');
+        // Show pending task from the UI
+        this._afterTokenReceived(data, cb);
+      },
+      data => {
+        this.logger.error(
+          'ShapeShift: GET Access Token: ERROR ' +
+            data.status +
+            '. ' +
+            this.getErrorsAsString(data.error)
+        );
+        return cb(data.error);
+      }
+    );
+  }
+
+  private _afterTokenReceived(data, cb) {
+    if (data && data.access_token && data.refresh_token) {
+      this.persistenceProvider.setShapeshiftToken(
+        this.credentials.NETWORK,
+        data.access_token
+      );
+      this.persistenceProvider.setShapeshiftRefreshToken(
+        this.credentials.NETWORK,
+        data.refresh_token
+      );
+      this.homeIntegrationsProvider.updateLink('shapeshift', data.access_token); // Name, Token
+      return cb(null, data.access_token);
+    } else {
+      return cb('Could not get the access token');
     }
   }
+
+  private _refreshToken(refreshToken, cb) {
+    let url = this.credentials.HOST + '/oauth/token';
+    let data = {
+      grant_type: 'refresh_token',
+      client_id: this.credentials.CLIENT_ID,
+      client_secret: this.credentials.CLIENT_SECRET,
+      redirect_uri: this.credentials.REDIRECT_URI,
+      refresh_token: refreshToken
+    };
+    const headers = new HttpHeaders({
+      'Content-Type': 'application/json',
+      Accept: 'application/json'
+    });
+
+    this.http.post(url, data, { headers }).subscribe(
+      data => {
+        this.logger.info('ShapeShift: Refresh Access Token SUCCESS');
+        this._afterTokenReceived(data, cb);
+      },
+      data => {
+        this.logger.error(
+          'ShapeShift: Refresh Access Token ERROR ' +
+            data.status +
+            '. ' +
+            this.getErrorsAsString(data.error)
+        );
+        return cb(data.error);
+      }
+    );
+  }
+
+  public getErrorsAsString(data): string {
```

Comment:
ShapeShift returns an object as error: { error: 'code', error_description: 'description' }
We should simplify this method.